lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -134,6 +144,95 @@ class ProgramLoweringState {
   DenseMap<Operation *, ComponentLoweringState> compStates;
 };
 
+//===----------------------------------------------------------------------===//
+// Partial lowering infrastructure
+//===----------------------------------------------------------------------===//
+
+/// Base class for partial lowering passes. A partial lowering pass
+/// modifies the root operation in place, but does not replace the root
+/// operation.
+/// The RewritePatternType template parameter allows for using both
+/// OpRewritePattern (default) or OpInterfaceRewritePattern.
+template <class OpType,
+          template <class> class RewritePatternType = OpRewritePattern>
+class PartialLoweringPattern : public RewritePatternType<OpType> {
+public:
+  using RewritePatternType<OpType>::RewritePatternType;
+  PartialLoweringPattern(MLIRContext *ctx, LogicalResult &resRef)
+      : RewritePatternType<OpType>(ctx), partialPatternRes(resRef) {}
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    rewriter.updateRootInPlace(
+        op, [&] { partialPatternRes = partiallyLower(op, rewriter); });
+    return partialPatternRes;
+  }
+
+  virtual LogicalResult partiallyLower(OpType op,
+                                       PatternRewriter &rewriter) const = 0;
+
+private:
+  LogicalResult &partialPatternRes;
+};
+
+//===----------------------------------------------------------------------===//
+// Partial lowering patterns
+//===----------------------------------------------------------------------===//
+
+/// FuncOpPartialLoweringPatterns are patterns which intend to match on FuncOps
+/// and then perform their own walking of the IR. FuncOpPartialLoweringPatterns
+/// have direct access to the ComponentLoweringState for the corresponding
+/// component of the matched FuncOp.
+class FuncOpPartialLoweringPattern
+    : public PartialLoweringPattern<mlir::FuncOp> {
+public:
+  FuncOpPartialLoweringPattern(MLIRContext *context, LogicalResult &resRef,
+                               FuncMapping &_funcMap, ProgramLoweringState &pls)
+      : PartialLoweringPattern(context, resRef), funcMap(_funcMap), pls(pls) {}
+
+  LogicalResult partiallyLower(mlir::FuncOp funcOp,
+                               PatternRewriter &rewriter) const override final {
+    // Initialize the component op references if a calyx::ComponentOp has been
+    // created for the matched funcOp.
+    auto it = funcMap.find(funcOp);
+    if (it != funcMap.end()) {
+      compOp = &it->second;
+      compLoweringState = &pls.compLoweringState(*comp());
+    }
+
+    return PartiallyLowerFuncToComp(funcOp, rewriter);
+  }
+
+  // Returns the component operation associated with the currently executing
+  // partial lowering.
+  calyx::ComponentOp *comp() const {
+    assert(compOp != nullptr);
+    return compOp;
+  }
```

Comment:
For my understanding, can you explain when and where this will be used?

Also, please add error message to this and the below `assert(compLoweringState != nullptr)`.

Reply:
An example use-case will be when a lowering pattern needs to i.e. create a new group within the component. In this case, the insertion point needs to be set to the `Wires` region of the component.
The `FuncOpPartialLoweringPattern` always matches a `mlir::FuncOp` which will have a 1:1 mapping between a corresponding `ComponentOp` that is being lowered - that we can conveniently access through these functions.

Reply:
Makes sense! Maybe a better name here, such as `getTargetedComponent` or `getLoweredComponent` or whatever may make this more clear when being used. If I just saw `comp()` being used in the future it wouldn't give me much information.